File: CPP/src/MultiscaleBlockLBP.cpp
#include "LBPBase.hpp"
#include <vector>
namespace LBP
{

namespace detail
{

MultiscaleBlockLBP::MultiscaleBlockLBP(unsigned int scale, unsigned int angle) : _scale(scale), _angle(angle)
{
	if ((_scale % 2) == 0)
		_scale += 1;
	_angle = _angle % 8;
}
MultiscaleBlockLBP::~MultiscaleBlockLBP()
{

}
bool MultiscaleBlockLBP::run_mutable(const cv::Mat& src, cv::Mat& dst)
{
	int rows = src.rows;
	int cols = src.cols;
	
	// 根据尺度计算网络信息
	int cellsz = _scale / 3;
	int cellar = cellsz * cellsz;
	int offset = cellsz / 2;

	int xCntRes = cols % cellsz;
	int xCntNum = cols / cellsz + (xCntRes == 0 ? 0 : 1);

	int yCntRes = rows % cellsz;
	int yCntNum = rows / cellsz + (yCntRes == 0 ? 0 : 1);

	if (xCntNum < 3 || yCntNum < 3 || src.type() != 0)
		return false;

	

	int xCurrentBlock(0), yCurrentBlock(0); // 当前block的左上顶点
	typedef unsigned char T;
	
	// get average pixel ahead
	std::vector<std::vector<double>> avg_pixel;
	avg_pixel.resize(xCntNum);
	std::vector<std::vector<T>> lbp_pixel;
	lbp_pixel.resize(xCntNum);
	for (int m = 0; m < xCntNum; ++m)
	{
		avg_pixel[m].resize(yCntNum, 0.0);
		lbp_pixel[m].resize(yCntNum, 0);
		for (int n = 0; n < yCntNum; ++n)
		{
			xCurrentBlock = cellsz * m;
			yCurrentBlock = cellsz * n;
			if (m == (xCntNum - 1) || n == (xCntNum - 1))
			{
				int totalpix = 0;
				int rowsL = yCntRes == 0 ? cellsz : yCntRes;
				int colsL = xCntRes == 0 ? cellsz : xCntRes;
				int resar = rowsL * colsL;
				for (int i = 0; i < rowsL; ++i)
				{
					for (int j = 0; j < xCntRes; ++j)
					{
						T val = src.at<T>(i + yCurrentBlock, j + xCurrentBlock);
						totalpix += (int)(val);
					}
				}
				avg_pixel[m][n] = (double)totalpix / (double)resar;
			}
			else
			{
				int totalpix = 0;
				for (int i = 0; i < cellsz; ++i)
				{
					for (int j = 0; j < cellsz; ++j)
					{
						T val = src.at<T>(i + yCurrentBlock, j + xCurrentBlock);
						totalpix += (int)(val);
					}
				}
				avg_pixel[m][n] = (double)totalpix / (double)cellar;
			}
			
		}
	}
	
	// cal true offset
	const int neighbor = 8;
	int angle = _angle;
	int direction_x[] = { -1, -1, 0, 1, 1, 1, 0, -1 };
	int direction_y[] = { 0, 1, 1, 1, 0, -1, -1, -1 };
	int idx(0);
	int dx(0), dy(0);
	
	for (int k = 0; k < neighbor; ++k)
	{
		idx = (k + angle) % neighbor;
		dx = direction_x[idx];
		dy = direction_y[idx];
		
		for (int m = 1; m < xCntNum - 1; ++m)
		{
			for (int n = 1; n < yCntNum - 1; ++n)
			{
				T center = static_cast<T>(avg_pixel[m][n]);
				T middle = static_cast<T>(avg_pixel[m + dx][n + dy]);
				T signal = center < middle ? 1 : 0;
				lbp_pixel[m][n] |= signal << (neighbor - k - 1);
			}
		}
	}
	dst = cv::Mat(cellsz * (yCntNum - 2), cellsz * (xCntNum - 2), CV_8UC1);
	int cur_rows(0);
	int cur_cols(0);
	for (int m = 1; m < xCntNum - 1; ++m)
	{
		for (int n = 1; n < yCntNum - 1; ++n)
		{
			cur_rows = (n - 1) * cellsz;
			cur_cols = (m - 1) * cellsz;
			for (int cur_r = 0; cur_r < cellsz; ++cur_r)
			{
				for (int cur_c = 0; cur_c < cellsz; ++cur_c)
				{
					dst.at<T>(cur_r + cur_rows, cur_c + cur_cols) = lbp_pixel[m][n];
				}
			}
		}
	}

	return true;
}


}
}

File: RUST/src/MultiscaleBlockLBP.rs
use super::LBPBase::IBase;
extern crate image;
use image::{GrayImage, Luma, Pixel};

pub struct MultiscaleBlockLBP
{
    _scale : u32,
    _angle : u32,
}
impl MultiscaleBlockLBP
{
    pub fn new(scale : u32, angle : u32) -> MultiscaleBlockLBP
    {
        return MultiscaleBlockLBP{_scale : scale, _angle : angle};
    }
}
impl IBase for MultiscaleBlockLBP
{
    fn run_mutable(&self, src : &GrayImage) -> GrayImage
    {
        let (rows, cols) : (u32, u32) = src.dimensions();
        let cellsz : u32 = self._scale / 3;
        let cellar : u32 = cellsz * cellsz;
        let offset : u32 = cellsz / 2;

        let xCntRes = rows % cellsz;
        let xCntNum = if xCntRes > 0 {rows / cellsz + 1} else { rows / cellsz };

        let yCntRes = cols % cellsz;
        let yCntNum = if yCntRes > 0 {cols / cellsz + 1} else { cols / cellsz };

        if xCntNum < 3 || yCntNum < 3 || cellsz == 0
        {
            let mut dst = GrayImage::new(rows, cols);
            return dst;
        }

        let (mut xCurrentBlock, mut yCurrentBlock) : (u32, u32) = (0, 0);

        let mut avg_pixs : Vec<Vec<f32>> = vec![vec![0.0; yCntNum as usize]; xCntNum as usize];
        let mut lbp_pixs : Vec<Vec<u8>> = vec![vec![0u8; yCntNum as usize]; xCntNum as usize];

        for m in 0 .. xCntNum
        {
            for n in 0 .. yCntNum
            {
                xCurrentBlock = cellsz * m;
                yCurrentBlock = cellsz * n;
                if m == (xCntNum - 1) || n == (yCntNum - 1)
                {
                    let mut totalpix : u32 = 0;
                    let rowsL = if xCntRes == 0 {cellsz} else {xCntRes};
			        let colsL = if yCntRes == 0 {cellsz} else {yCntRes};
                    let resar = rowsL * colsL;
                    for i in 0 .. rowsL
                    {
                        for j in 0 .. colsL
                        {
                            let val : u8 = src.get_pixel(xCurrentBlock + i, yCurrentBlock + j).channels()[0];
                            totalpix += val as u32;
                        }
                    }
                    if resar > 0 
                    {
                        avg_pixs[m as usize][n as usize] = (totalpix as f32) / (resar as f32);}
                    }
                    else
                    {
                        let mut totalpix : u32 = 0;
                        for i in 0 .. cellsz
                        {
                            for j in 0 .. cellsz
                            {
                                let val : u8 = src.get_pixel(xCurrentBlock + i, yCurrentBlock + j).channels()[0];
                                totalpix += val as u32;
                            }
                        }
				        avg_pixs[m as usize][n as usize] = (totalpix as f32)/ (cellar as f32);
                    }
                }
            }

            let neighbor : u32 = 8;
            let angle : u32 = self._angle;
            let direction_x : [i32; 8] = [-1, -1, 0, 1, 1, 1, 0, -1 ];
            let direction_y : [i32; 8] = [0, 1, 1, 1, 0, -1, -1, -1 ];
            let mut idx : u32 = 0;
            let (mut dx, mut dy) : (i32, i32) = (0, 0);
            for k in 0 .. neighbor
            {
                idx = (k + angle) % neighbor;
                dx = direction_x[idx as usize];
                dy = direction_y[idx as usize];

                for m in 1 .. (xCntNum - 1)
                {
                    for n in 1 .. (yCntNum - 1)
                    {
                        let center : u8 = avg_pixs[m as usize][n as usize] as u8;
                        let px : usize = ((m as i32) + dx) as usize;
                        let py : usize = ((n as i32) + dy) as usize;
                        let edges : u8 = avg_pixs[px][py] as u8;
                        let signal : u8 = if center < edges { 1u8 } else { 0u8 };
                        lbp_pixs[m as usize][n as usize]  |= signal << (neighbor - k - 1);
                    }
                }
            }
        let mut dst = GrayImage::new(cellsz * (xCntNum - 2), cellsz * (yCntNum - 2));
        let (mut cur_rows, mut cur_cols) : (u32, u32) = (0, 0);
        for m in 1 .. (xCntNum - 1)
        {
            for n in 1 .. (yCntNum - 1)
            {
                let cur_rows = (m - 1) * cellsz;
                let cur_cols = (n - 1) * cellsz;
                for cur_r in 0 .. cellsz
                {
                    for cur_c in 0 .. cellsz
                    {
                        let dst_p = dst.get_pixel_mut(cur_r + cur_rows, cur_c + cur_cols).channels_mut();
                        dst_p[0] = lbp_pixs[m as usize][n as usize] as u8; 
                    }
                }
            }
        }
        return dst;
    }
}